scipy/signal/_signaltools.py

```diff
@@ -2452,6 +2456,269 @@ def hilbert2(x, N=None):
+    ...                        sharez=ax00)
+    >>> ax00.set_title("Analytic Signal")
+    >>> ax00.set(xlim=(0, 1), ylim=(-ma, ma), zlim=(-ma, ma))
+    >>> ax01.set_title("Real-valued Signal")
+    >>> for z_, e_, ax_ in zip((z, z.real), (e_a, e_r), (ax00, ax01)):
+    ...     ax_.set(xlabel="Time $t$", ylabel="Real Amp. $x(t)$",
+    ...             zlabel="Imag. Amp. $y(t)$")
+    ...     ax_.plot(t, z_.real, 'C0-', zs=-ma, zdir='z', alpha=0.5, label="Real")
+    ...     ax_.plot_surface(E2d_t, e_*E2_amp, -ma*E2d_1, color='C1', alpha=0.25)
+    ...     ax_.plot(t, z_.imag, 'C0-', zs=+ma, zdir='y', alpha=0.5, label="Imag.")
+    ...     ax_.plot_surface(E2d_t, ma*E2d_1, e_*E2_amp, color='C1', alpha=0.25)
+    ...     ax_.plot(t, z_.real, z_.imag, 'C0-', label="Signal")
+    ...     ax_.plot_surface(E3d_t, e_*np.cos(E3d_phi), e_*np.sin(E3d_phi),
+    ...                      color='C1', alpha=0.5, shade=True, label="Envelope")
+    ...     ax_.view_init(elev=22.7, azim=-114.3)
+    >>> fg0.subplots_adjust(left=0.08, right=0.97, wspace=0.15)
+    >>> plt.show()
+
+    """
+    if not (-z.ndim <= axis < z.ndim):
+        raise ValueError(f"Invalid parameter {axis=} for {z.shape=}!")
+    if not (z.shape[axis] > 0):
+        raise ValueError(f"z.shape[axis] not > 0 for {z.shape=}, {axis=}!")
+    if len(bp_in) != 2 or not all((isinstance(b_, int) or b_ is None) for b_ in bp_in):
+        raise ValueError(f"{bp_in=} isn't a 2-tuple of type (int | None, int | None)!")
+    if not ((isinstance(n_out, int) and 0 < n_out) or n_out is None):
+        raise ValueError(f"{n_out=} is not a positive integer or None!")
+    if residual not in ('lowpass', 'all', None):
+        raise ValueError(f"{residual=} not in ['lowpass', 'all', None]!")
+
+    n = z.shape[axis]  # number of time samples of input
+    n_out = n if n_out is None else n_out
+    fak = n_out / n  # scaling factor for resampling
+
+    bp = slice(bp_in[0] if bp_in[0] is not None else -(n//2),
+               bp_in[1] if bp_in[1] is not None else (n+1)//2)
+    if not (-n//2 <= bp.start < bp.stop <= (n+1)//2):
+        raise ValueError("`-n//2 <= bp_in[0] < bp_in[1] <= (n+1)//2` does not hold " +
+                         f"for n={z.shape[axis]=} and {bp_in=}!")
+
+    # moving active axis to end allows to use `...` for indexing:
+    z = np.moveaxis(z, axis, -1)
+
+    if np.iscomplexobj(z):
+        Z = sp_fft.fft(z)
+    else:  # avoid calculating negative frequency bins for real signals:
+        Z = np.zeros_like(z, dtype=sp_fft.rfft(z.flat[:1]).dtype)
+        Z[..., :n//2 + 1] = sp_fft.rfft(z)
+        if bp.start > 0:  # make signal analytic within bp_in band:
+            Z[..., bp] *= 2
+        elif bp.stop > 0:
+            Z[..., 1:bp.stop] *= 2
+    if not (bp.start <= 0 < bp.stop):  # envelope is invariant to freq. shifts.
+        z_bb = sp_fft.ifft(Z[..., bp], n=n_out) * fak  # baseband signal
```

Comment:
Why do we need to shift it otherwise?

Reply:
To save a little bit of computational effort by shortening the `ifft` input